### ISSUE TITLE
[Java Client][REQUIREMENT] Support format message with simple argument, which means argument format（type and style） can be specified in the argument placeholder.

### ISSUE THREAD

**Issue by huihuiw01** [vmware/singleton#1424]
**Is your feature request related to a problem? Please describe.**
       This is a requirement from product‘s developer.

       Currently we support when there are arguments in the message string, before the arguments are replaced by real value, real value must be localized(when it's a message string)/formatted(when it's a L2 argument) by calling L3/L2 API.
e.g. 
               translationMessage.getMessage(locale, component,key, args[])

       When the message corresponding to the key is: 
               At {1} on {1}, there was {2} on planet {0}.
       
       The English arguments value is:
Object[] arguments = {
                7,
                new Date(timestamp),
                "a disturbance in the Force"
        };

       When the locale is 'fr', arguments value should be in 'fr' style, which means string argument should be 'fr' translation by calling L3 API and number&date argument should be in 'fr' format by calling L2 API.

        In this case, not only the developer need call Singleton client's API twice, but also the message is difficult for translator to translate.
       

       Now product's developer has a number/date argument, he doesn't want to call L2 API first to format the argument, then call L3 API to get the message. He hopes he can get message including number/date argument with once L3 API call by specifying the format of argument in the message string. In another word, he wants the message string can be defined as below: (can specify the argument type and style)

                  At {1,date,shortTime} on {1,date,longDate}, there was {2} on planet {0,number}.

      After calling L3 API, he can get the message with arguments formatted according to the specified type(e.g. date) and style(e.g. longDate).

**Describe the solution you'd like**
Currently we plan to support number and date type first.

supported argType = "number" | "date"

supported argStyle for "number" type = "currency" | "percent" 
Note:
1. argStyle is optional, when it's not specified, the number will be formatted in decimal format

supported argStyle for "date" type = "shortDate" | "mediumDate" | "longDate" | "fullDate" | "shortTime" | "mediumTime" | "longTime" | "fullTime" | "short" | "medium" | "long" | "full"
Note:
1. The format results of "short" | "medium" | "long" | "full" style are the combination of date and time.
2. When argStyle for "date" is not specified, the date will be formatted in "medium" style

**Describe alternatives you've considered**

**Additional context**

**Comment by nick-stephen** [comment]
Any news on this?

For reference here's a section from the conversation on #g11n-vip-assist:

Thanks for getting back to me.  I still don’t understand the formatting structure of the ‘source’ string (or its localized version). The javadoc doesn’t cover that. You mention CLDR which I looked up, but I don’t live in the i18n world and I don’t understand if the format of this string meets this CLDR standard.

You mention using a CLDR library - I googled and found a couple such as ‘https://github.com/Squarespace/cldr-engine-java’ which says its’s compatible with https://phensley.github.io/cldr-engine/docs/en/doc-messageformatting which references this ICU MessageFormat class:

https://unicode-org.github.io/icu-docs/apidoc/released/icu4j/com/ibm/icu/text/MessageFormat.html

Which DOES specify the format of the patterns and their interpretation, finally. I’m not sure if this is what you implement.
But this one states how to describe how you want to format ‘Date’ arguments - finally!
